/*
 * Beautiful Capi generates beautiful C API wrappers for your C++ classes
 * Copyright (C) 2015 Petr Petrovich Petrov
 *
 * This file is part of Beautiful Capi.
 *
 * Beautiful Capi is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Beautiful Capi is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Beautiful Capi.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

/*
 * WARNING: This file was automatically generated by Beautiful Capi!
 * Do not edit this file! Please edit the source API description.
 */

#ifndef EXCEPTION_BADARGUMENT_DEFINITION_INCLUDED
#define EXCEPTION_BADARGUMENT_DEFINITION_INCLUDED

#include "Exception/BadArgumentDecl.h"
#include "Exception/Generic.h"
#include "Exception/common/check_and_throw_exception.h"

#ifdef __cplusplus

inline Exception::BadArgument::BadArgument() : Exception::Generic(Exception::Generic::force_creating_from_raw_pointer, 0, false)
{
    beautiful_capi_exception_exception_info_t exception_info;
    void* result(exception_bad_argument_new(&exception_info));
    beautiful_capi_Exception::check_and_throw_exception(exception_info.code, exception_info.object_pointer);
    SetObject(result);
}

inline const char* Exception::BadArgument::GetArgumentName() const
{
    return exception_bad_argument_get_argument_name(this->GetRawPointer());
}

inline Exception::BadArgument::BadArgument(const BadArgument& other) : Exception::Generic(Exception::Generic::force_creating_from_raw_pointer, 0, false)
{
    if (other.mObject)
    {
        beautiful_capi_exception_exception_info_t exception_info;
        void* result(exception_bad_argument_copy(&exception_info, other.mObject));
        beautiful_capi_Exception::check_and_throw_exception(exception_info.code, exception_info.object_pointer);
        SetObject(result);
    }
    else
    {
        SetObject(0);
    }
}

#ifdef EXCEPTION_CPP_COMPILER_HAS_RVALUE_REFERENCES
inline Exception::BadArgument::BadArgument(BadArgument&& other) : Exception::Generic(std::move(other))
{
    mObject = other.mObject;
    other.mObject = 0;
}
#endif /* EXCEPTION_CPP_COMPILER_HAS_RVALUE_REFERENCES */

inline Exception::BadArgument::BadArgument(Exception::BadArgument::ECreateFromRawPointer, void *object_pointer, bool copy_object) : Exception::Generic(Exception::Generic::force_creating_from_raw_pointer, 0, false)
{
    if (object_pointer && copy_object)
    {
        beautiful_capi_exception_exception_info_t exception_info;
        void* result(exception_bad_argument_copy(&exception_info, object_pointer));
        beautiful_capi_Exception::check_and_throw_exception(exception_info.code, exception_info.object_pointer);
        SetObject(result);
    }
    else
    {
        SetObject(object_pointer);
    }
}

inline Exception::BadArgument::~BadArgument()
{
    if (mObject && Exception::Generic::mObject)
    {
        exception_bad_argument_delete(mObject);
        SetObject(0);
    }
}

inline Exception::BadArgument& Exception::BadArgument::operator=(const Exception::BadArgument& other)
{
    if (this != &other)
    {
        if (mObject && Exception::Generic::mObject)
        {
            exception_bad_argument_delete(mObject);
            SetObject(0);
        }
        if (other.mObject)
        {
            beautiful_capi_exception_exception_info_t exception_info;
            void* result(exception_bad_argument_copy(&exception_info, other.mObject));
            beautiful_capi_Exception::check_and_throw_exception(exception_info.code, exception_info.object_pointer);
            SetObject(result);
        }
        else
        {
            SetObject(0);
        }
    }
    return *this;
}

#ifdef EXCEPTION_CPP_COMPILER_HAS_RVALUE_REFERENCES
inline Exception::BadArgument& Exception::BadArgument::operator=(Exception::BadArgument&& other)
{
    if (this != &other)
    {
        if (mObject && Exception::Generic::mObject)
        {
            exception_bad_argument_delete(mObject);
            SetObject(0);
        }
        Exception::Generic::operator=(std::move(other));
        mObject = other.mObject;
        other.mObject = 0;
    }
    return *this;
}
#endif /* EXCEPTION_CPP_COMPILER_HAS_RVALUE_REFERENCES */

inline Exception::BadArgument Exception::BadArgument::Null()
{
    return Exception::BadArgument(Exception::BadArgument::force_creating_from_raw_pointer, 0, false);
}

inline bool Exception::BadArgument::IsNull() const
{
    return !mObject;
}

inline bool Exception::BadArgument::IsNotNull() const
{
    return mObject != 0;
}

inline bool Exception::BadArgument::operator!() const
{
    return !mObject;
}

inline void* Exception::BadArgument::Detach()
{
    void* result = mObject;
    SetObject(0);
    return result;
}

inline void* Exception::BadArgument::GetRawPointer() const
{
    return mObject;
}

inline void Exception::BadArgument::SetObject(void* object_pointer)
{
    mObject = object_pointer;
    if (mObject)
    {
        Exception::Generic::SetObject(exception_bad_argument_cast_to_base(mObject));
    }
    else
    {
        Exception::Generic::SetObject(0);
    }
}

#endif /* __cplusplus */

#endif /* EXCEPTION_BADARGUMENT_DEFINITION_INCLUDED */

